keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+- Resource overcommitment, fractional resource requirements, native compute resource (i.e., cpu and memory) with special metadata requirements, and group compute resources. They are out of our current scope.
+
+## Components
+### ResourceClass API
+During the initial phase, we propose to start with the following ResourceClass API spec that defines the basic ResourceClass name to the underlying node level compute resource plus metadata matching. We can extend this API to better support cluster resources and group resources in the following phases of development. However, this document will mostly focus on the design to support initial phase requirements.
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec ResourceClassSpec
+        // +optional
+}
+
+type ResourceClassSpec struct {
+        // defines general resource property matching constraints.
+        // e.g.: zone in { us-west1-b, us-west1-c }; type: k80
+        MetadataRequirements metav1.LabelSelector
+        // used to compare preference of two matching ResourceClasses
+        // The purpose to introduce this field is explained more later
+        Priority int
+}
+```
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Kind"
+          operator: "In"
+          values:
+            - "nvidia-gpu"
+        - key: "memory"
+          operator: "GtEq"
+          values:
+            - "30G"
+
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  containers:
+    - name: example-container
+      resources:
+        limits:
+          nvidia.high.mem: 2
+
+```
+Above resource class will select all the nvidia-gpus which have memory greater
+than and equal to 30 GB.
+
+YAML example 2:
+```yaml
+kind: ResourceClass
+metadata:
+  name: fast.nic
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Kind"
+          operator: "In"
+          values:
+            - "nic"
+        - key: "speed"
+          operator: "GtEq"
+          values:
+            - "40GBPS"
+
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  containers:
+    - name: example-container
+      resources:
+        limits:
+          fast.nic: 1
+```
+Above resource class will select all the NICs with speed greater than equal to
+40 GBPS.
+
+### Kubelet Extension
+On node level, extended resources can be exported automatically by a third-party plugin through the Device Plugin API. We propose to extend the current Device Plugin API that allows device plugins to send to the Kubelet per-device properties during device listing. Exporting device properties at per-device level instead of per-resource level allows a device plugin to manage devices with heterogeneous properties.
```

Comment:
some real, current use-case for Vish: I was recently wondering ho to upgrade the SRIOV Device Plugin to fit the needs of TelCo applications better

currently, the SRIOV device plugin exposes how many VFs are available from a host. but it does not expose which PF those VFs belong to
it also does not expose which NUMA socket ID those VFs belong to

However, these information would be required by certain applications for:
- SCTP multi-homing: this is a mandatory requirement for some control plane mobile network applications. Such application will require two, physically separated NICs. Simply asking 2 VFs from a host will not guarantee that, but if Device Manager would store the PFs of the VFs too, the user could explicitly ask two VFs to be allocated to it from different PFs
- similarly with NUMA: when NUMA manager is introduced (keeping my fingers crossed...), it will query Device Manager for NUMA hints. If the VFs of a node would be grouped by their NUMA socket ID, then Device Manager could provide an educated hint to NUMA manager. Then, during Allocate() DM could pick a VF from the aligned NUMA group

Reply:
one other, totally unrelated thing came into my mind: when SmartNICs will come into the picture, they will contain FPGAs too.
Some smart NICs will contain a an OVS control plane software embedded. Some smart nics will contain storage control plane softwares embedded into them. Some smart NICs will contain some special crypto related instruction sets embedded into them.
Some smart NICs will contain all three somewhere in the future, based on what I have heard.

To be more generic: having an "FPGA special resource" available on the node does not tell anything, because that FPGA can contain sooooo many things! It depends on its programming

Reply:
@vishh ^

Reply:
1. NUMA/device topology is totally orthogonal to this proposal. I agree we need to solve that and I can help with that. 
2. Excuse my ignorance, what is a PF and what's it's relationship with VF?
3. The FPGA use cases sound interesting, but I'd like to understand why node level granularity isn't adequate enough. Dynamically re-programming FPGA and thereby changing scheduling attributes within a single node seems like a niche use case to me to warrant adding a large new API.

Reply:
1: great to hear :)
2: VF, is the short for "Virtual Function". PF is "Physical Function". One NIC is a PF, which can be divided into multiple "VF"s via th SRIOV kernel feature.
Effectively a "VF" provides direct access to the "PF" -shaving of the delays coming from overlay networking- while also dividing up the capabilities of the PF between the VFs.
E.g. let's we divide one PF into 8 VFs, then we can share the same NIC between 8 containers, and all 8 containers will get direct access to the NIC's HW resources.
3: As you say everything can be solved with labels and taints, but the aim here would be solving it in an efficient way.
My two cents is that "Pod explicitly asks for resource made available on a Node by a DP" is a way more explicit and binding contract compared to "Pod wants to be schedule to a Node called "IHaveThatResources,Really,BelieveMe".
If we would like to handle FPGAs in the future as resources, I foresee that we might want to differentiate between them.
Ofc same could also achieved by creating a different DP per "FPGA-program", but that approach would seem kind of wasteful to me(?)
Anyway, I was just trying to came-up with use-cases from my field, I'm not saying one problem can't have multiple solutions!

Reply:
@vishh 
> Dynamically re-programming FPGA and thereby changing scheduling attributes within a single node seems like a niche use case to me to warrant adding a large new API.

I'd argue that dynamic reprogramming is one of "must to have" usage scenarios for FPGA. It's essential for "Orchestration programmed Accelerated Function" workflow, where bitstream should be programmed by orchestration machinery before running a workflow that consumes it.

Reply:
For FPGA bitstream programming use case, my feeling from the discussions in 
 https://github.com/kubernetes/kubernetes/pull/61775 is that the missing part of a long-term solution is a field in ContainerSpec through which we can pass workload level special configs to set up a resource. Here are some differences I think we may draw the line between this config setting and the ResourceClass config setting: a) Consumer of such container level configs is Kubelet or Kubelet level plugins, not scheduler, so such config settings shouldn't affect pod scheduling decisions. b) Such container level settings shouldn't affect resource management policies such as quota. Given such differences, perhaps we should consider to make AllocatedComputeResources a struct (perhaps should also rename it somehow) so that we can add a field later that can be used for this purpose.

Reply:
@jiayingz Can you explain how to use this approach in more details? Something similar to @vikaschoudhary16  explanation [here](https://github.com/kubernetes/kubernetes/pull/61775#issuecomment-403831327) may be?

Reply:
Basically we will add https://github.com/kubernetes/kubernetes/pull/58388 in DPAPI. Then the general workflow is:
1. A device plugin sends Kubelet list of devices and properties for each device.
2. Kubelet exports device properties through the new ComputeResources field, e.g.:
```
Node: {
  Status: {
    ComputeResources: [
    {
	Name: nvidia-gpu-a7e3f2d0
	ResourceName: nvidia.com/gpu
	Properties: {
		Type: K80
                Memory: 12G
	}
	Devices: { "nvidia0", "nvidia1" }
        AllocatableDevices: { "nvidia0", "nvidia1" }
    }, {
	Name: nvidia-gpu-b3d00dff
	ResourceName: nvidia.com/gpu
	Properties: {
		Type: P100
                Memory: 16G
	}
	Devices: { "nvidia2", "nvidia3" }
        AllocatableDevices: { "nvidia2", "nvidia3" }
    }]
  }
}
```
3. Cluster operator creates a ResourceClass, e.g.:
```
kind: ResourceClass
metadata:
  name: nvidia.high.mem
spec:
  resourceName: "nvidia.com/gpu"
  labelSelector:
    - matchExpressions:
        - key: "memory"
          operator: "GtEq"
          values:
            - "15G"
```
4. Workload requests the resource through the resource name:
```
kind: Pod
metadata:
  name: example-pod
spec:
  containers:
    - name: example-container
      resources:
        limits:
          nvidia.high.mem: 2
```
5. Scheduler schedule the pod on the node. Note scheduler updates its cached ComputeResource to ResourceClass matching as soon as it sees the ResourceClass creation update. During Pod scheduling, it simply uses the cached matching information to do the basic resource accounting during PodFitsNodeResource evaluation. After scheduler chooses a node for the pod, it populates AllocatedComputeResources field in Pod:
```
AllocatedComputeResources {
    nvidia.high.mem: []{
        { ComputeResourceName: nvidia-gpu-b3d00dff, Count: 2 },
   }
```
6. Pod reaches Kubelet and Kubelet allocates nvidia2 and nvidia3 to the Pod, and sends device plugin an allocate call.

Reply:
@jiayingz Thank you for detailed explanations!

I'm still having hard time understanding how this could be used to specify FPGA bitstreams for reprogramming. 
Unlike GPU properties from your example bitstreams are not properties. They're rather content that FPGA devices can be programmed with. Using them as properties doesn't look like reasonable solution to me.

You've mentioned workload specific configs as a possible long term solution for FPGA programming. Can you explain in more details how this can be used and give an example? Thank you.

Reply:
Sorry @bart0sh my above example workflow is based on what is covered in the current proposal. For FPGA bitstream programming, as I mentioned in https://github.com/kubernetes/community/pull/2265#discussion_r202769458 I feel the missing part is a field in ContainerSpec through which we can pass workload level special configs to set up a resource. I.e., in the step 4 of the above workflow example, we can instead have the following Pod spec with 
```
Workload requests the resource through the resource name:
kind: Pod
metadata:
  name: example-pod
spec:
  containers:
    - name: example-container
      resources:
        limits:
          fpga-bitstream: 2
      allocatedComputeResources []{
          "fpga-bitstream": {
              resourceRequestParameters: []{
                  "fpga-function": "bitstream-100"
              }
              // AllocatedResourceList is left out for scheduler to populate during step 5. 
          }
```
However, it will require more thinking on whether we want to encode this resource request information in allocatedComputeResources, or propose a separate field, or just start with some special pod annotation field.

Reply:
@jiayingz thank you for the explanations!

From the first look it still looks unnecessary complex and harder to use than pod annotations in my opinion. It also looks a bit artificial to use FPGA bitstream as a resource class as it's not a device property. It's an external content that can be programmed into the device and change its properties. Until it's programmed it doesn't "belong" to the device.

Reply:
My concern for pod annotation is it is meant for temporary experiment, and we still to figure out long term solution. Even for short term experiment, I think the general policy now is that we should use alpha field instead of pod annotation. I tend to agree that encoding the resource config request info in allocatedComputeResources is perhaps not the right way. If we do need to support such use cases for multiple kinds of devices, I feel the longer term solution would perhaps require some kind of API change on ContainerSpec. On the other hand, if FPGA is the only use case so far that requires such workload level dynamic resource config, I wonder whether it can be solved with FPGA specific API, e.g., a FPGA daemon on the host can listen on a unix socket through which user application can send bitstream programming request?